[JSC] Handle multiple Temporal calendar annotations

ptomato opened the pull request:
#### 09fe0b34ffb6b1f0b547f7f6e4b9e3f4f0da54ba
<pre>
[JSC] Handle multiple Temporal calendar annotations
<a href="https://bugs.webkit.org/show_bug.cgi?id=223166">https://bugs.webkit.org/show_bug.cgi?id=223166</a>

Reviewed by Justin Michaud.

This is part of a normative change to the Temporal proposal:
<a href="https://github.com/tc39/proposal-temporal/pull/2397">https://github.com/tc39/proposal-temporal/pull/2397</a>
This change allows parsing multiple annotations in Temporal strings. If
there are multiple `u-ca` calendar annotations, the first one is used.
But if there are multiple calendars and any have the `!` flag, the
string is rejected because that&apos;s inconsistent.

This makes a lot of test262 tests newly pass and they can be removed
from the skip list. Some other tests under the same heading still fail
because PlainDate and PlainDateTime don&apos;t yet support calendars, so
recategorize them under the &quot;Depends on calendar support&quot; heading.

* JSTests/stress/temporal-instant.js:
* JSTests/test262/config.yaml:
* Source/JavaScriptCore/runtime/ISO8601.cpp:
(JSC::ISO8601::parseOneCalendar):
(JSC::ISO8601::parseCalendar):
(JSC::ISO8601::parseCalendarTime):
(JSC::ISO8601::parseCalendarDateTime):

Canonical link: <a href="https://commits.webkit.org/293742@main">https://commits.webkit.org/293742@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b486223802c1ce5422d6318df4003450bf4bf309

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/99790 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/19435 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/9709 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/104915 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/50373 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/19740 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/27871 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/75954 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/33048 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/102797 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/15035 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/90096 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/56320 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/14843 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/8083 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/49740 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/92452 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/84767 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/8168 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/107276 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/98398 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/26901 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/19640 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/84911 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/27263 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/86301 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/84436 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21440 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/29109 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/6818 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/20709 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/26839 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/32051 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/122022 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/26650 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/34067 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/29967 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/28214 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

justinmichaud commented on Source/JavaScriptCore/runtime/ISO8601.cpp:
```diff
@@ -813,12 +813,11 @@ static std::optional<TimeZoneRecord> parseTimeZone(StringParsingBuffer<Character
 
-    if (!canBeCalendar(buffer))
-        return std::nullopt;
+    ASSERT(canBeCalendar(buffer));
```
I wonder if this should be a release assert, or perhaps we can keep the existing nullopt return case, just in case this can somehow be reached.

ptomato replied:
In that case, maybe it's easiest to drop the assertion and keep a nullopt return. In this function it's trivial to prove the assertion (`parseOneCalendar` is called only in a `while(canBeCalendar(buffer))` loop in `parseCalendar`) but the other site below is less straightforward to prove.

justinmichaud commented on Source/JavaScriptCore/runtime/ISO8601.cpp:
```diff
@@ -866,12 +865,7 @@ static std::optional<CalendarRecord> parseCalendar(StringParsingBuffer<Character
-        if (isCritical)
-            return std::nullopt;
-        return CalendarRecord { /* isCritical = */ false, /* isUnknown = */ true, { } };
+        return CalendarRecord { isCritical, /* isUnknown = */ true, { } };
```
Can we get an enum for isUnknown? In general, I like seeing named enums instead of bools, although I don't think this should block this patch.

ptomato replied:
Sure, I can add that in the next patch. Would you want `isCritical` to be an enum as well?

ptomato replied:
→ https://github.com/WebKit/WebKit/pull/44132

justinmichaud commented on Source/JavaScriptCore/runtime/ISO8601.cpp:
```diff
@@ -921,6 +915,42 @@ static std::optional<CalendarRecord> parseCalendar(StringParsingBuffer<Character
+    //  Annotations :::
+    //      Annotation Annotations[opt]
+
+    ASSERT(canBeCalendar(buffer));
```
ditto

catamorphism committed:
[JSC] Handle multiple Temporal calendar annotations
https://bugs.webkit.org/show_bug.cgi?id=223166

Reviewed by Justin Michaud.

This is part of a normative change to the Temporal proposal:
https://github.com/tc39/proposal-temporal/pull/2397
This change allows parsing multiple annotations in Temporal strings. If
there are multiple `u-ca` calendar annotations, the first one is used.
But if there are multiple calendars and any have the `!` flag, the
string is rejected because that's inconsistent.

This makes a lot of test262 tests newly pass and they can be removed
from the skip list. Some other tests under the same heading still fail
because PlainDate and PlainDateTime don't yet support calendars, so
recategorize them under the "Depends on calendar support" heading.

* JSTests/stress/temporal-instant.js:
* JSTests/test262/config.yaml:
* Source/JavaScriptCore/runtime/ISO8601.cpp:
(JSC::ISO8601::parseOneCalendar):
(JSC::ISO8601::parseCalendar):
(JSC::ISO8601::parseCalendarTime):
(JSC::ISO8601::parseCalendarDateTime):

Canonical link: https://commits.webkit.org/293742@main
